File: src/main/java/cn/wjybxx/dson/DsonTimestamp.java
package cn.wjybxx.dson;

import cn.wjybxx.dson.types.OffsetTimestamp;

import javax.annotation.Nonnull;

/**
 * @author wjybxx
 * date - 2023/6/17
 */
public class DsonTimestamp extends DsonValue {

    private final OffsetTimestamp value;

    public DsonTimestamp(OffsetTimestamp value) {
        this.value = value;
    }

    public OffsetTimestamp getValue() {
        return value;
    }

    @Nonnull
    @Override
    public DsonType getDsonType() {
        return DsonType.TIMESTAMP;
    }

    //region equals

    @Override
    public boolean equals(Object o) {
        if (this == o) return true;
        if (o == null || getClass() != o.getClass()) return false;

        DsonTimestamp that = (DsonTimestamp) o;

        return value.equals(that.value);
    }

    @Override
    public int hashCode() {
        return value.hashCode();
    }

    // endregion

    @Override
    public String toString() {
        return "DsonTimestamp{" +
                "value=" + value +
                '}';
    }
}

File: csharp/Dson/src/DsonTimestamp.cs
﻿#region LICENSE

//  Copyright 2023-2024 wjybxx(845740757@qq.com)
// 
//  Licensed under the Apache License, Version 2.0 (the "License");
//  you may not use this file except in compliance with the License.
//  You may obtain a copy of the License at
// 
//      http://www.apache.org/licenses/LICENSE-2.0
// 
//  Unless required by applicable law or agreed to iBn writing, software
//  distributed under the License is distributed on an "AS IS" BASIS,
//  WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
//  See the License for the specific language governing permissions and
//  limitations under the License.

#endregion

using System;
using Wjybxx.Dson.Types;

#pragma warning disable CS1591
namespace Wjybxx.Dson;

/// <summary>
/// Dson时间戳类型
/// </summary>
public class DsonTimestamp : DsonValue, IEquatable<DsonTimestamp>
{
    private readonly OffsetTimestamp _value;

    public DsonTimestamp(OffsetTimestamp value) {
        _value = value;
    }

    public override DsonType DsonType => DsonType.Timestamp;
    public OffsetTimestamp Value => _value;

    #region equals

    public bool Equals(DsonTimestamp? other) {
        if (ReferenceEquals(null, other)) return false;
        if (ReferenceEquals(this, other)) return true;
        return _value.Equals(other._value);
    }

    public override bool Equals(object? obj) {
        if (ReferenceEquals(null, obj)) return false;
        if (ReferenceEquals(this, obj)) return true;
        if (obj.GetType() != this.GetType()) return false;
        return Equals((DsonTimestamp)obj);
    }

    public override int GetHashCode() {
        return _value.GetHashCode();
    }

    public static bool operator ==(DsonTimestamp? left, DsonTimestamp? right) {
        return Equals(left, right);
    }

    public static bool operator !=(DsonTimestamp? left, DsonTimestamp? right) {
        return !Equals(left, right);
    }

    #endregion

    public override string ToString() {
        return $"{nameof(DsonType)}: {DsonType}, {nameof(_value)}: {_value}";
    }
}